<a href="https://colab.research.google.com/github/adhang/data-science-digitalskola/blob/update/08.%20Advanced%20Pandas%20Dataframe/Learn%20-%20Advanced%20Pandas%20Dataframe.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Advanced Dataframe
Author: Adhang Muntaha Muhammad

[![LinkedIn](https://img.shields.io/badge/linkedin-0077B5?style=for-the-badge&logo=linkedin&logoColor=white&link=https://www.linkedin.com/in/adhangmuntaha/)](https://www.linkedin.com/in/adhangmuntaha/)
[![GitHub](https://img.shields.io/badge/github-121011?style=for-the-badge&logo=github&logoColor=white&link=https://github.com/adhang)](https://github.com/adhang)
[![Kaggle](https://img.shields.io/badge/kaggle-20BEFF?style=for-the-badge&logo=kaggle&logoColor=white&link=https://www.kaggle.com/adhang)](https://www.kaggle.com/adhang)
[![Tableau](https://img.shields.io/badge/tableau-E97627?style=for-the-badge&logo=tableau&logoColor=white&link=https://public.tableau.com/app/profile/adhang)](https://public.tableau.com/app/profile/adhang)
___
**Contents**
- Indexing Dataframe
- Dropping Columns
- Joining Dataframes
- Contatenating Dataframes
- Appending Dataframes
- Pivot Table
- Melting Dataframes
- Lambda Function on Dataframes

# Importing Libraries

In [2]:
import pandas as pd

# Reading Dataset
For this notebook, I will use my mentor's dataset from GitHub. You can check his works [here](https://github.com/densaiko).

In [12]:
file_path = 'https://raw.githubusercontent.com/densaiko/data_science_learning/main/dataset/insurance.csv'

data = pd.read_csv(file_path)
data.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

# Indexing Dataframe

## Default Index
An index is used to identify a row/ record. The default index starts from 0 to n (determined by the total rows).

For example, let's see the dataset size using `shape`.

In [4]:
data.shape

(1338, 7)

As we can see, there are 1338 rows and 7 columns. Let's see the first 5 rows.

In [5]:
data.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

From this output, we know that the index is started from 0 to 4. How about the last index?

In [6]:
data.tail()

age     sex    bmi  children smoker     region     charges
1333   50    male  30.97         3     no  northwest  10600.5483
1334   18  female  31.92         0     no  northeast   2205.9808
1335   18  female  36.85         0     no  southeast   1629.8335
1336   21  female  25.80         0     no  southwest   2007.9450
1337   61  female  29.07         0    yes  northwest  29141.3603

Now, we know the last 5 rows. The last row has 1337 as the index. So, an index ranging from 0 to 1337 means it has 1338 rows.

## Column as Index
Can we change the index using a column? Yes, absolutely. We can use `set_index()` method to do this.

Let's say we will use the `sex` column as the index.

In [20]:
data.set_index('sex')

age     bmi  children smoker     region      charges
sex                                                         
female   19  27.900         0    yes  southwest  16884.92400
male     18  33.770         1     no  southeast   1725.55230
male     28  33.000         3     no  southeast   4449.46200
male     33  22.705         0     no  northwest  21984.47061
male     32  28.880         0     no  northwest   3866.85520
...     ...     ...       ...    ...        ...          ...
male     50  30.970         3     no  northwest  10600.54830
female   18  31.920         0     no  northeast   2205.98080
female   18  36.850         0     no  southeast   1629.83350
female   21  25.800         0     no  southwest   2007.94500
female   61  29.070         0    yes  northwest  29141.36030

[1338 rows x 6 columns]

Note, it's just for displaying the result. If you want to save it, use `inplace=True` or assign it to a variable (whether it's a new variable or the same variable). Like this:

```
# assign to the original dataframe
data.set_index('sex', inplace=True)

# assign to a new variable
new_data = data.set_index('sex')

# assign to the same variable
data = data.set_index('sex')
```



## Multiple Index
We can set multiple columns as the index. It will create a multi-index.

In [18]:
data.set_index(['sex','smoker'])

age     bmi  children     region      charges
sex    smoker                                               
female yes      19  27.900         0  southwest  16884.92400
male   no       18  33.770         1  southeast   1725.55230
       no       28  33.000         3  southeast   4449.46200
       no       33  22.705         0  northwest  21984.47061
       no       32  28.880         0  northwest   3866.85520
...            ...     ...       ...        ...          ...
       no       50  30.970         3  northwest  10600.54830
female no       18  31.920         0  northeast   2205.98080
       no       18  36.850         0  southeast   1629.83350
       no       21  25.800         0  southwest   2007.94500
       yes      61  29.070         0  northwest  29141.36030

[1338 rows x 5 columns]

Here, the `sex` and `smoker` are set as the index.

## Reset Index
Let's say, our dataframe has an index that confusing (or not ordered). We can reset the index using `reset_index()`.

To demonstrate this, I will create a new dataframe using random sampling. So, if you re-run this notebook, you may get a different result.

In [19]:
data_5 = data.sample(n=5)
data_5

age     sex    bmi  children smoker     region     charges
833    58    male  34.39         0     no  northwest  11743.9341
1065   42  female  25.30         1     no  southwest   7045.4990
976    48    male  40.15         0     no  southeast   7804.1605
721    53    male  36.60         3     no  southwest  11264.5410
785    35  female  27.70         3     no  southwest   6414.1780

This dataframe contains i

# Dropping Columns

In [ ]:
drop_1 = data.drop('charges', axis=1)
drop_1.head()

age     sex     bmi  children smoker     region
0   19  female  27.900         0    yes  southwest
1   18    male  33.770         1     no  southeast
2   28    male  33.000         3     no  southeast
3   33    male  22.705         0     no  northwest
4   32    male  28.880         0     no  northwest

In [ ]:
drop_2 = data.drop(['age','sex'], axis=1)
drop_2.head()

bmi  children smoker     region      charges
0  27.900         0    yes  southwest  16884.92400
1  33.770         1     no  southeast   1725.55230
2  33.000         3     no  southeast   4449.46200
3  22.705         0     no  northwest  21984.47061
4  28.880         0     no  northwest   3866.85520

In [ ]:
drop_row = data.drop(1, axis=0)
drop_row.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520
5   31  female  25.740         0     no  southeast   3756.62160

In [ ]:
data_sex = data.set_index('sex')
data_sex.head()

age     bmi  children smoker     region      charges
sex                                                         
female   19  27.900         0    yes  southwest  16884.92400
male     18  33.770         1     no  southeast   1725.55230
male     28  33.000         3     no  southeast   4449.46200
male     33  22.705         0     no  northwest  21984.47061
male     32  28.880         0     no  northwest   3866.85520

In [ ]:
data_sex_drop = data_sex.drop('female', axis=0)
data_sex_drop.head()

age     bmi  children smoker     region      charges
sex                                                       
male   18  33.770         1     no  southeast   1725.55230
male   28  33.000         3     no  southeast   4449.46200
male   33  22.705         0     no  northwest  21984.47061
male   32  28.880         0     no  northwest   3866.85520
male   37  29.830         2     no  northeast   6406.41070

# Joining Dataframes

In [ ]:
data_left = data.sample(n=5)
data_right = data.sample(n=10)

display(data_left.head())
display(data_right.head())

age     sex     bmi  children smoker     region      charges
28     23    male  17.385         1     no  northwest   2775.19215
82     22    male  37.620         1    yes  southeast  37165.16380
283    55  female  32.395         1     no  northeast  11879.10405
761    23    male  35.200         1     no  southwest   2416.95500
1044   55    male  35.245         1     no  northeast  11394.06555

age     sex    bmi  children smoker     region      charges
1174   29    male  32.11         2     no  northwest   4433.91590
576    22    male  26.84         0     no  southeast   1664.99960
377    24    male  40.15         0    yes  southeast  38126.24650
1332   52  female  44.70         3     no  southwest  11411.68500
1223   20  female  24.42         0    yes  southeast  26125.67477

In [ ]:
data_left.join(data_right, lsuffix='_first', rsuffix='_second')

age_first sex_first  ...  region_second  charges_second
28           23      male  ...            NaN             NaN
82           22      male  ...            NaN             NaN
283          55    female  ...            NaN             NaN
761          23      male  ...            NaN             NaN
1044         55      male  ...            NaN             NaN

[5 rows x 14 columns]

# Concatenating Dataframes

In [ ]:
pd.concat([data_left, data_right], axis=1)

age     sex     bmi children  ... children smoker     region      charges
28    23.0    male  17.385      1.0  ...      NaN    NaN        NaN          NaN
82    22.0    male  37.620      1.0  ...      NaN    NaN        NaN          NaN
283   55.0  female  32.395      1.0  ...      NaN    NaN        NaN          NaN
377    NaN     NaN     NaN      NaN  ...      0.0    yes  southeast  38126.24650
576    NaN     NaN     NaN      NaN  ...      0.0     no  southeast   1664.99960
683    NaN     NaN     NaN      NaN  ...      0.0     no  northwest   9863.47180
761   23.0    male  35.200      1.0  ...      NaN    NaN        NaN          NaN
1044  55.0    male  35.245      1.0  ...      NaN    NaN        NaN          NaN
1122   NaN     NaN     NaN      NaN  ...      3.0    yes  northwest  46661.44240
1169   NaN     NaN     NaN      NaN  ...      1.0     no  northwest   6112.35295
1174   NaN     NaN     NaN      NaN  ...      2.0     no  northwest   4433.91590
1210   NaN     NaN     NaN      NaN  ...      1.0     no  northwest   5373.36425
1223   NaN     NaN     NaN      NaN  ...      0.0    yes  southeast  26125.67477
1263   NaN     NaN     NaN      NaN  ...      1.0     no  southwest   7337.74800
1332   NaN     NaN     NaN      NaN  ...      3.0     no  southwest  11411.68500

[15 rows x 14 columns]

In [ ]:
pd.concat([data_left, data_right], axis=0)

age     sex     bmi  children smoker     region      charges
28     23    male  17.385         1     no  northwest   2775.19215
82     22    male  37.620         1    yes  southeast  37165.16380
283    55  female  32.395         1     no  northeast  11879.10405
761    23    male  35.200         1     no  southwest   2416.95500
1044   55    male  35.245         1     no  northeast  11394.06555
1174   29    male  32.110         2     no  northwest   4433.91590
576    22    male  26.840         0     no  southeast   1664.99960
377    24    male  40.150         0    yes  southeast  38126.24650
1332   52  female  44.700         3     no  southwest  11411.68500
1223   20  female  24.420         0    yes  southeast  26125.67477
683    53    male  24.320         0     no  northwest   9863.47180
1263   43  female  29.900         1     no  southwest   7337.74800
1169   37  female  34.105         1     no  northwest   6112.35295
1210   36    male  30.875         1     no  northwest   5373.36425
1122   53  female  36.860         3    yes  northwest  46661.44240

# Appending Dataframes

In [ ]:
data_left.append(data_right)

age     sex     bmi  children smoker     region      charges
28     23    male  17.385         1     no  northwest   2775.19215
82     22    male  37.620         1    yes  southeast  37165.16380
283    55  female  32.395         1     no  northeast  11879.10405
761    23    male  35.200         1     no  southwest   2416.95500
1044   55    male  35.245         1     no  northeast  11394.06555
1174   29    male  32.110         2     no  northwest   4433.91590
576    22    male  26.840         0     no  southeast   1664.99960
377    24    male  40.150         0    yes  southeast  38126.24650
1332   52  female  44.700         3     no  southwest  11411.68500
1223   20  female  24.420         0    yes  southeast  26125.67477
683    53    male  24.320         0     no  northwest   9863.47180
1263   43  female  29.900         1     no  southwest   7337.74800
1169   37  female  34.105         1     no  northwest   6112.35295
1210   36    male  30.875         1     no  northwest   5373.36425
1122   53  female  36.860         3    yes  northwest  46661.44240

# Pivot Table

In [ ]:
data.set_index(['sex','smoker'])

age     bmi  children     region      charges
sex    smoker                                               
female yes      19  27.900         0  southwest  16884.92400
male   no       18  33.770         1  southeast   1725.55230
       no       28  33.000         3  southeast   4449.46200
       no       33  22.705         0  northwest  21984.47061
       no       32  28.880         0  northwest   3866.85520
...            ...     ...       ...        ...          ...
       no       50  30.970         3  northwest  10600.54830
female no       18  31.920         0  northeast   2205.98080
       no       18  36.850         0  southeast   1629.83350
       no       21  25.800         0  southwest   2007.94500
       yes      61  29.070         0  northwest  29141.36030

[1338 rows x 5 columns]

In [ ]:
data.groupby(['sex','smoker','region']).age.mean().unstack()

region         northeast  northwest  southeast  southwest
sex    smoker                                            
female no      39.840909  39.755556  39.071942  40.099291
       yes     38.724138  38.827586  39.250000  37.047619
male   no      39.216000  38.568182  38.261194  40.277778
       yes     37.868421  39.827586  40.054545  35.567568

In [ ]:
pd.pivot_table(data, values='age', index=['sex','smoker'], columns='region', aggfunc='mean')

region         northeast  northwest  southeast  southwest
sex    smoker                                            
female no      39.840909  39.755556  39.071942  40.099291
       yes     38.724138  38.827586  39.250000  37.047619
male   no      39.216000  38.568182  38.261194  40.277778
       yes     37.868421  39.827586  40.054545  35.567568

In [ ]:
pd.pivot_table(data, values='age', index=['sex','smoker'], columns='region', aggfunc=['mean','max'])

mean                        ...       max                    
region         northeast  northwest  southeast  ... northwest southeast southwest
sex    smoker                                   ...                              
female no      39.840909  39.755556  39.071942  ...        64        64        64
       yes     38.724138  38.827586  39.250000  ...        64        64        64
male   no      39.216000  38.568182  38.261194  ...        64        64        64
       yes     37.868421  39.827586  40.054545  ...        62        64        61

[4 rows x 8 columns]